File: rxbinding-design/src/main/java/com/jakewharton/rxbinding2/support/design/widget/SwipeDismissBehaviorObservable.java
package com.jakewharton.rxbinding2.support.design.widget;

import android.support.design.widget.CoordinatorLayout.LayoutParams;
import android.support.design.widget.SwipeDismissBehavior;
import android.support.design.widget.SwipeDismissBehavior.OnDismissListener;
import android.view.View;
import io.reactivex.Observable;
import io.reactivex.Observer;
import io.reactivex.android.MainThreadDisposable;

import static com.jakewharton.rxbinding2.internal.Preconditions.checkMainThread;

final class SwipeDismissBehaviorObservable extends Observable<View> {
  private final View view;

  SwipeDismissBehaviorObservable(View view) {
    this.view = view;
  }

  @Override protected void subscribeActual(Observer<? super View> observer) {
    if (!checkMainThread(observer)) {
      return;
    }
    if (!(view.getLayoutParams() instanceof LayoutParams)) {
      throw new IllegalArgumentException("The view is not in a Coordinator Layout.");
    }
    LayoutParams params = (LayoutParams) view.getLayoutParams();
    final SwipeDismissBehavior behavior = (SwipeDismissBehavior) params.getBehavior();
    if (behavior == null) {
      throw new IllegalStateException("There's no behavior set on this view.");
    }
    Listener listener = new Listener(behavior, observer);
    observer.onSubscribe(listener);
    behavior.setListener(listener);
  }

  static final class Listener extends MainThreadDisposable implements OnDismissListener {
    private final SwipeDismissBehavior swipeDismissBehavior;
    private final Observer<? super View> observer;

    Listener(SwipeDismissBehavior swipeDismissBehavior, Observer<? super View> observer) {
      this.swipeDismissBehavior = swipeDismissBehavior;
      this.observer = observer;
    }

    @Override public void onDismiss(View view) {
      if (!isDisposed()) {
        observer.onNext(view);
      }
    }

    @Override public void onDragStateChanged(int state) {
    }

    @Override protected void onDispose() {
      swipeDismissBehavior.setListener(null);
    }
  }
}

File: rxbinding-material/src/main/java/com/jakewharton/rxbinding3/material/SwipeDismissBehaviorObservable.kt
@file:JvmName("RxSwipeDismissBehavior")
@file:JvmMultifileClass

package com.jakewharton.rxbinding3.material

import android.view.View
import androidx.annotation.CheckResult
import androidx.coordinatorlayout.widget.CoordinatorLayout.LayoutParams
import com.google.android.material.behavior.SwipeDismissBehavior
import com.google.android.material.behavior.SwipeDismissBehavior.OnDismissListener
import com.jakewharton.rxbinding2.internal.Preconditions.checkMainThread
import io.reactivex.Observable
import io.reactivex.Observer
import io.reactivex.android.MainThreadDisposable

/**
 * Create an observable which emits the dismiss events from `view` on
 * [SwipeDismissBehavior].
 *
 * *Warning:* The created observable keeps a strong reference to `view`. Unsubscribe
 * to free this reference.
 */
@CheckResult
fun View.dismisses(): Observable<View> {
  return SwipeDismissBehaviorObservable(this)
}

private class SwipeDismissBehaviorObservable(
  private val view: View
) : Observable<View>() {

  override fun subscribeActual(observer: Observer<in View>) {
    if (!checkMainThread(observer)) {
      return
    }
    val params = view.layoutParams as? LayoutParams
        ?: throw IllegalArgumentException("The view is not in a Coordinator Layout.")
    val behavior = params.behavior as SwipeDismissBehavior<*>?
        ?: throw IllegalStateException("There's no behavior set on this view.")

    val listener = Listener(behavior, observer)
    observer.onSubscribe(listener)
    behavior.setListener(listener)
  }

  private class Listener(
    private val swipeDismissBehavior: SwipeDismissBehavior<*>,
    private val observer: Observer<in View>
  ) : MainThreadDisposable(), OnDismissListener {

    override fun onDismiss(view: View) {
      if (!isDisposed) {
        observer.onNext(view)
      }
    }

    override fun onDragStateChanged(state: Int) {}

    override fun onDispose() {
      swipeDismissBehavior.setListener(null)
    }
  }
}
